From fafa54b22fa3743ea94b9614aefe84f2daffd409 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Thu, 8 Dec 2016 13:40:48 +0800
Subject: [PATCH 1372/5242] MLK-13607-4 video: mxsfb: change enable/disable
 controller logic.

commit  afdec8bc2868c7568de23949dc9551f97d93f56c from
https://source.codeaurora.org/external/imx/linux-imx.git

Change the enable/disable controller logic to support mipi dsi
display on imx7ulp-evk board.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 657d3a052581d517bb6c9cd978b04c4a3194c350)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxsfb.c |   15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index fca1d46..e4339e6 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -546,6 +546,13 @@ static void mxsfb_enable_controller(struct fb_info *fb_info)
 		}
 	}
 
+	if (host->dispdrv && host->dispdrv->drv->enable) {
+		ret = host->dispdrv->drv->enable(host->dispdrv, fb_info);
+		if (ret < 0)
+			dev_err(&host->pdev->dev, "failed to enable "
+				"dispdrv:%s\n", host->dispdrv->drv->name);
+	}
+
 	/* the pixel clock should be disabled before
 	 * trying to set its clock rate successfully.
 	 */
@@ -586,12 +593,6 @@ static void mxsfb_enable_controller(struct fb_info *fb_info)
 
 	host->enabled = 1;
 
-	if (host->dispdrv && host->dispdrv->drv->enable) {
-		ret = host->dispdrv->drv->enable(host->dispdrv, fb_info);
-		if (ret < 0)
-			dev_err(&host->pdev->dev, "failed to enable "
-				"dispdrv:%s\n", host->dispdrv->drv->name);
-	}
 }
 
 static void mxsfb_disable_controller(struct fb_info *fb_info)
@@ -603,6 +604,8 @@ static void mxsfb_disable_controller(struct fb_info *fb_info)
 
 	dev_dbg(&host->pdev->dev, "%s\n", __func__);
 
+	writel(CTRL_RUN, host->base + LCDC_CTRL + REG_CLR);
+
 	if (host->dispdrv && host->dispdrv->drv->disable)
 		host->dispdrv->drv->disable(host->dispdrv, fb_info);
 
-- 
1.7.9.5

